06-building-a-program-start-to-finish/solving-the-problem/output/InvertedIndex.js

```diff
@@ -69,90 +69,100 @@ const InvertedIndex = class {
-    if (this.indexedBookTitles[bookname]) {
-      this.indexedBookTitles[bookname].push(title);
-    } else {
-      this.indexedBookTitles[bookname] = [];
-      this.indexedBookTitles[bookname].push(title);
-    }
+    return text.replace(/[^\w\s]/gi, '')
+      .replace(/\s+/g, ' ')
+      .trim()
+      .toLowerCase();
   }
 
   /**
    * Creates a word array for each document in a book
+   * @param {String} bookName - name of a book to index
    * @param {Object} book - documents in a book
    * @return {Array.<Object>} bookContents - words in each document
    */
-  createsArray(bookname, book) {
+  createsArray(bookName, book) {
     const bookContents = [];
     Object.keys(book).map((documentPosition) => {
       const mergedTitleAndText = `${book[documentPosition].title} 
       ${book[documentPosition].text}`;
       bookContents.push(InvertedIndex.tokenize(mergedTitleAndText).split(' '));
-      this.setBookTitles(bookname, book[documentPosition].title);
+      this.setBookTitles(bookName, book[documentPosition].title);
+      return this;
     });
     return bookContents;
   }
 
   /**
+   * Sets title(s) for a book in an instance variable - indexedBookTitles
+   * @param {String} bookName - name of a book being indexed
+   * @param {String} title - a titlee in the book being indexed
+   * @returns {Void} indexedBookTitles - instance variable
+   */
+  setBookTitles(bookName, title) {
+    if (this.indexedBookTitles[bookName]) {
+      this.indexedBookTitles[bookName].push(title);
+    } else {
+      this.indexedBookTitles[bookName] = [];
+      this.indexedBookTitles[bookName].push(title);
+    }
+  }
+
+  /**
    * Creates the index for specified book
-   * @param {string} bookname - Name of book to be indexed
+   * @param {string} bookName - Name of book to be indexed
    * @param {object} book - document words in specified book
    * @return {Promise.<iDexMapper>} - indexes of specified book
    */
-  createIndex(bookname, book) {
+  createIndex(bookName, book) {
     const tokenIndex = {};
-    this.numberOfDocuments[bookname] = [];
-    const bookContents = this.createsArray(bookname, book);
+    this.numberOfDocuments[bookName] = [];
+    const bookContents = this.createsArray(bookName, book);
     return new Promise((resolve) => {
       bookContents.map((eachdocument, documentPosition) => {
         const documentPositionToInt = parseInt(documentPosition, 10);
-        this.numberOfDocuments[bookname].push(documentPositionToInt);
-        eachdocument.map((eachWord) => {
+        this.numberOfDocuments[bookName].push(documentPositionToInt);
+        return eachdocument.map((eachWord) => {
           if (tokenIndex[eachWord]) {
             if (tokenIndex[eachWord].indexOf(documentPositionToInt) === -1) {
               tokenIndex[eachWord].push(documentPositionToInt);
             }
           } else {
             tokenIndex[eachWord] = [documentPositionToInt];
           }
+          return this;
```

Comment:
By default, functional methods such as map return an implicit 'this'. Which is used to support method chaining.
In this scenario, I am returning this `tokenIndex[eachWord]`, thereby making it possible to chain other methods on the map method. I have however, changed it to explicitly return `tokenIndex[eachWord]`